src/NUnitFramework/framework/Attributes/RetryAttribute.cs

```diff
@@ -91,10 +91,10 @@ public override TestResult Execute(TestExecutionContext context)
 
                 while (count-- > 0)
                 {
-                    context.CurrentResult = innerCommand.Execute(context);
+                    innerCommand.ExecuteAndSetCurrentResult(context);
 
-                    if (context.CurrentResult.ResultState != ResultState.Failure)
-                        break;
+                    if (context.CurrentResult.ResultState.Status != TestStatus.Failed) break;
+                    if (context.CurrentResult.ResultState.Label == ResultState.Error.Label) break;
```

Comment:
This code that separately tests status and label makes no sense to me, especially now that you fixed the problem whereby we were using object equality. It peeks into the implementation of a test outcome (`ResultState`) and deigns to parse the internals, which can easily change. It also changes behavior when the `Label` is neither empty nor "Error".

I think we might want to further encapsulate test outcome rather than making use of the internals.

Reply:
I think making use of the internal label is justified because that is exactly what the test is specced to do. The code transparently shows that. It's *not* specced to react differently if the site fails to match, so I could not consider comparing the `ResultState` directly.

Reply:
But the `Site`, at this point, is guaranteed to be `FailureSite.Test`. I think you are trying to make this more general than it really is - we only do retry on individual test cases.

Reply:
That's not obvious to me. How could I have figured out by myself that site is guaranteed to be Test?

Reply:
I suppose you only could have by a very wide review of the code, much wider than you would want to do for the purpose of this issue. So it's understandable that you wouldn't know that.

OTOH, I don't think it's a really great pattern for every member of a team to have to figure out things on their own. We need to have ways to tell each other the stuff we know and to rely on one another for information. In this case, I thought it was something you already knew.

Which is not to say that the internal design of ResultState couldn't be better. Clearly it could. But I think that the improvement should be in applying the Open Closed Principle so it works better without revealing more of it's internals.